File: src/test/java/com/github/kayjamlang/tests/expressions/NotExpressionTest.java
package com.github.kayjamlang.tests.expressions;

import com.github.kayjamlang.core.expressions.Expression;
import com.github.kayjamlang.core.KayJamLexer;
import com.github.kayjamlang.core.KayJamParser;
import com.github.kayjamlang.core.expressions.ValueExpression;
import com.github.kayjamlang.core.expressions.NegationExpression;
import org.junit.BeforeClass;
import org.junit.Test;

import static org.junit.Assert.*;

public class NotExpressionTest {

    private static KayJamParser parser;

    @BeforeClass
    public static void prepare(){
        parser = new KayJamParser(new KayJamLexer("!true"));
    }

    @Test
    public void test() throws Exception {
        Expression expression = parser.readExpression();

        assertNotNull(expression);
        assertSame(NegationExpression.class, expression.getClass());

        NegationExpression notExpression = (NegationExpression) expression;
        assertSame(ValueExpression.class, notExpression.expression.getClass());

        ValueExpression constant = (ValueExpression) notExpression.expression;
        assertEquals(true, constant.value);
    }
}

File: src/test/scala/com/github/kayjamlang/tests/expressions/NotExpressionTest.scala
package com.github.kayjamlang.tests.expressions

import com.github.kayjamlang.core.{KayJamLexer, KayJamParser}
import com.github.kayjamlang.core.expressions.{NegationExpression, ValueExpression}
import org.junit.Assert._
import org.junit.{BeforeClass, Test}

object NotExpressionTest {
  private var parser: KayJamParser = null

  @BeforeClass def prepare(): Unit = {
    parser = new KayJamParser(new KayJamLexer("!true"))
  }
}

class NotExpressionTest {
  @Test
  @throws[Exception]
  def test(): Unit = {
    val expression = NotExpressionTest.parser.readExpression
    assertNotNull(expression)
    assertSame(classOf[NegationExpression], expression.getClass)
    val notExpression = expression.asInstanceOf[NegationExpression]
    assertSame(classOf[ValueExpression], notExpression.expression.getClass)
    val constant = notExpression.expression.asInstanceOf[ValueExpression]
    assertEquals(true, constant.value)
  }
}